rustdoc: cleanups relating to allocations

nnethercote opened the pull request:
These commits generally clean up the code a bit and also reduce allocation rates a bit.

r? @camelid

nnethercote committed:
Avoid `Box` in `href_relative_parts`.

This reverts part of #91948, going back to returning a
`UrlPartsBuilder`. It makes the code simpler, and also avoids some
allocations.

nnethercote committed:
Rename some methods.

Most of the methods returning `impl Display` have `print` in their name.
This commit renames a few that didn't follow that convention.

nnethercote committed:
Simplify `make_href`.

It never fails, so it doesn't need to return `Result`. And the
`ItemType` in the result is just a copy of the one passed in via the
`shortty` arg, so it can also be removed.

nnethercote committed:
Make `{Type,Path}::generics` return iterators.

Instead of a `Vec`, to avoid some allocations.

nnethercote committed:
Avoid some unnecessary cloning.

nnethercote committed:
Simplify `make_href`.

Currently it is passed an `fqp` slice which it calls `to_vec` on and
returns. This is a bit odd. It's better to let the call site clone if
necessary. (One call site does, one does not).

nnethercote committed:
Refactor the end of `generate_item_def_id_path`.

To avoids the early return and duplication of `Ok((url_parts, shortty,
fqp))`.

nnethercote committed:
Streamline `TypeAliasPart::get`.

- `ret` only ever gets at most one entry, so it can be an `Option`
  instead of a `Vec`.
- Which means we can use `filter_map` instead of `flat_map`.
- Move `trait_` next to the `ret` assignment, which can only happen
  once.
- No need for `impls` to be a `Vec`, it can remain an iterator.
- Avoid `Result` when collecting `impls`.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15254018293/job/42904905413)
Build commit: 9656ba7f36cc1c5221657e832f653689644cb6b3 (`9656ba7f36cc1c5221657e832f653689644cb6b3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15254018293/job/42904905413"},"merge_sha":"9656ba7f36cc1c5221657e832f653689644cb6b3"} -->

rust-timer commented:
Queued 9656ba7f36cc1c5221657e832f653689644cb6b3 with parent 95a2212587f1b8500af1ac630322b0cb2e030c61, future [comparison URL](https://perf.rust-lang.org/compare.html?start=95a2212587f1b8500af1ac630322b0cb2e030c61&end=9656ba7f36cc1c5221657e832f653689644cb6b3).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~3.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (9656ba7f36cc1c5221657e832f653689644cb6b3): [comparison URL](https://perf.rust-lang.org/compare.html?start=95a2212587f1b8500af1ac630322b0cb2e030c61&end=9656ba7f36cc1c5221657e832f653689644cb6b3&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=95a2212587f1b8500af1ac630322b0cb2e030c61&end=9656ba7f36cc1c5221657e832f653689644cb6b3&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-2.9%, -0.2%] | 11    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.6% | [-2.9%, -0.2%] | 11    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=95a2212587f1b8500af1ac630322b0cb2e030c61&end=9656ba7f36cc1c5221657e832f653689644cb6b3&stat=max-rss)
<details>
<summary>Results (primary -1.2%, secondary -2.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.2%, -1.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-2.7%, -2.7%] | 1     |
| All ❌✅ (primary)                 | -1.2% | [-1.2%, -1.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=95a2212587f1b8500af1ac630322b0cb2e030c61&end=9656ba7f36cc1c5221657e832f653689644cb6b3&stat=cycles:u)
<details>
<summary>Results (primary -2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.8% | [-2.8%, -2.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.8% | [-2.8%, -2.8%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=95a2212587f1b8500af1ac630322b0cb2e030c61&end=9656ba7f36cc1c5221657e832f653689644cb6b3&stat=size:linked_artifact)
<details>
<summary>Results (primary -1.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.1%, -1.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.1% | [-1.1%, -1.1%] | 1     |
</details>

**Bootstrap**: 776.359s -> 777.253s (0.12%)
**Artifact size**: 366.28 MiB -> 366.23 MiB (-0.01%)

lolbinarycat reviewed (commented):
1 potential further optimization, 1 nit about readability, and 1 question about rationale.

yotamofek commented on src/librustdoc/html/render/write_shared.rs:
```diff
@@ -682,7 +680,7 @@ impl TypeAliasPart {
 
             let part = OrderedJson::array_sorted(
-                impls.iter().map(OrderedJson::serialize).collect::<Result<Vec<_>, _>>().unwrap(),
+                impls.map(OrderedJson::serialize).collect::<Result<Vec<_>, _>>().unwrap(),
```
Just FYI that this `collect` is also redundant,
this will work too:

```suggestion
                impls.map(|impl_| OrderedJson::serialize(impl_).unwrap()),
```
(though doesn't seem to affect perf on my local tests)

nnethercote replied:
Nice suggestion! I have applied it.

lolbinarycat commented on src/librustdoc/html/format.rs:
```diff
@@ -1419,7 +1415,7 @@ pub(crate) fn visibility_print_with_space(item: &clean::Item, cx: &Context<'_>)
                 // modified from `resolved_path()` to work with `DefPathData`
                 let last_name = path.data.last().unwrap().data.get_opt_name().unwrap();
-                let anchor = anchor(vis_did, last_name, cx);
+                let anchor = print_anchor(vis_did, last_name, cx);
```
yes, but do you think it would be worth it to move the main function body into the `fmt::from_fn` call? it should save at least an allocation, no?

nnethercote replied:
Not worth it from a perf perspective: this function doesn't show up as hot in profiles (I've been using profiles to guide all of this work) and I think it will only allocate in the `else` branch containing `s`.

But: worth it from a code cleanliness perspective, because it makes the code shorter and nicer. I've added a new commit doing it, thanks for the suggestion.

nnethercote commented on src/librustdoc/html/tests.rs:
```diff
@@ -1,51 +1,51 @@
-    // the symbols used are static, and no `Symbol::intern` calls occur.
-    assert_eq!(expected, href_relative_parts(&fqp, &relative_to_fqp).collect::<Vec<_>>());
+fn assert_relative_path(expected: &str, relative_to_fqp: &[Symbol], fqp: &[Symbol]) {
+    create_default_session_globals_then(|| {
```
Without it:
```
thread 'html::tests::href_relative_parts_root' panicked at /home/njn/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/scoped-tls-1.0.1/src/lib.rs:168:9:
cannot access a scoped thread local variable without calling `set` first
```
The `sym::*` symbols are preinterned, i.e. interned at session startup, and the test then causes them to be accessed. Which requires that a session has started.

nnethercote commented on src/librustdoc/html/render/write_shared.rs:
```diff
@@ -660,16 +653,21 @@ impl TypeAliasPart {
                             )
                             .to_string();
-                            ret.push(AliasSerializableImpl {
+                            let trait_ = impl_
+                                .inner_impl()
+                                .trait_
+                                .as_ref()
+                                .map(|trait_| format!("{:#}", trait_.print(cx)));
+                            ret = Some(AliasSerializableImpl {
```
I have added a comment explaining the "what", but not the "why" (because I don't know why).

camelid replied:
> disables html escaping of < and >

That's not really the core of it. The alternate format prints things (in this case, the trait's path) as plain text rather than as rich HTML. So as part of that, it doesn't use HTML escaping. But the main thing is that here we're generating plaintext rather than HTML.

I'm not exactly sure of the reason, though likely we're doing some kind of matching against the path in JS to know which aliased impl to load. Anyway, it's beside the point of this PR.

nnethercote committed:
Streamline `visibility_print_with_space`.

Moving the visibility stuff into the `from_fn` avoids the `Cow` and
makes the code a little shorter and simpler.

nnethercote commented:
BTW, this is a case where selecting "Show non-relevant results" in the perf results is worthwhile. There are a number of sub-threshold improvements that I think are real, because they are so consistent.

camelid reviewed (approved):
Looks great, thanks! Just one nit to clean up the iterator code.

nnethercote commented on src/librustdoc/html/format.rs:
```diff
@@ -1119,8 +1119,8 @@ impl clean::Impl {
                 if self.kind.is_fake_variadic()
-                    && let generics = ty.generics()
-                    && let &[inner_type] = generics.as_ref().map_or(&[][..], |v| &v[..])
+                    && let Some(mut generics) = ty.generics()
+                    && let (Some(inner_type), None) = (generics.next(), generics.next())
```
Nice, I didn't know about that method.

nnethercote committed:
Address review comments.

bors commented:
:pushpin: Commit 68f32169e59dd8bde8fd66cc4ffe98cd24ac09ca has been approved by `camelid`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=camelid 68f32169e59dd8bde8fd66cc4ffe98cd24ac09ca -->
<!-- homu: {"type":"Approved","sha":"68f32169e59dd8bde8fd66cc4ffe98cd24ac09ca","approver":"camelid","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15343080438/job/43183745465)
Approved by: camelid
Pushing e6152cdf5b31bd844a4cc1049433859d54863602 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"camelid","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/15343080438/job/43183745465"},"merge_sha":"e6152cdf5b31bd844a4cc1049433859d54863602"} -->

rust-timer commented:
Finished benchmarking commit (e6152cdf5b31bd844a4cc1049433859d54863602): [comparison URL](https://perf.rust-lang.org/compare.html?start=6de3a733122a82d9b3c3427c7ee16a1e1a022718&end=e6152cdf5b31bd844a4cc1049433859d54863602&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6de3a733122a82d9b3c3427c7ee16a1e1a022718&end=e6152cdf5b31bd844a4cc1049433859d54863602&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.6%, -0.2%] | 8     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-0.6%, -0.6%] | 1     |
| All ❌✅ (primary)                 | -0.4% | [-0.6%, -0.2%] | 8     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6de3a733122a82d9b3c3427c7ee16a1e1a022718&end=e6152cdf5b31bd844a4cc1049433859d54863602&stat=max-rss)
<details>
<summary>Results (primary -2.6%, secondary 0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.6%  | [0.4%, 0.9%]   | 6     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-2.1%, -0.4%] | 3     |
| All ❌✅ (primary)                 | -2.6% | [-2.6%, -2.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6de3a733122a82d9b3c3427c7ee16a1e1a022718&end=e6152cdf5b31bd844a4cc1049433859d54863602&stat=cycles:u)
<details>
<summary>Results (secondary 0.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.0%  | [0.5%, 3.0%]   | 8     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-0.7%, -0.7%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6de3a733122a82d9b3c3427c7ee16a1e1a022718&end=e6152cdf5b31bd844a4cc1049433859d54863602&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data
**Artifact size**: 370.26 MiB -> 370.19 MiB (-0.02%)
